lib/compat.php

```diff
@@ -304,6 +304,286 @@ function gutenberg_warn_classic_about_blocks() {
+		 *		$scale = $content_width / $block_attributes['editWidth'];
+		 *		$image_width = round( $block_attributes['width'] * $scale );
+		 *
+		 *		$image_dimensions = wp_constrain_dimensions( $image_file_width, $image_file_height, $image_width );
+		 *	}
+		 */
+	} elseif ( ! empty( $block_attributes['fileWidth'] ) && ! empty( $block_attributes['fileHeight'] ) ) {
+		$image_dimensions = array(
+			$block_attributes['fileWidth'],
+			$block_attributes['fileHeight'],
+		);
+	} else {
+		return false;
+	}
+
+	/*
+	 * Do not constrain images with "wide" and "full" alignment to the "large" image size.
+	 * TODO: To reduce (fix) the need for upscaling or using the "full" size images
+	 * add "xlarge" image size generated by default!
+	 */
+	if (
+		! empty( $image_meta['width'] ) &&
+		! empty( $block_attributes['fileWidth'] ) &&
+		( 'wide' === $block_attributes['align'] || 'full' === $block_attributes['align'] )
+	) {
+		$size_updated = false;
+
+		// Attempt to find the largest image size that may have been added by themes or plugins.
+		if ( ! empty( $image_meta['sizes'] ) ) {
+			foreach ( $image_meta['sizes'] as $size ) {
+				if ( $size['width'] > $image_dimensions[0] && wp_image_matches_ratio( $block_attributes['fileWidth'], $block_attributes['fileHeight'], $size['width'], $size['height'] ) ) {
+					$image_dimensions = array(
+						$size['width'],
+						$size['height'],
+					);
+
+					$size_updated = true;
+				}
+			}
+		}
+
+		if (
+			! $size_updated &&
+			$block_attributes['fileWidth'] < $image_meta['width'] &&
+			// Do not force site visitors to download HUGE images.
+			// Max 12 MP photo (that's still pretty arbitrary, may be over 3MB, consider reducing).
+			max( (int) $image_meta['width'], (int) $image_meta['height'] ) < 4300 &&
+			wp_image_matches_ratio( $block_attributes['fileWidth'], $block_attributes['fileHeight'], $image_meta['width'], $image_meta['height'] )
+		) {
+			$image_dimensions = array(
+				$image_meta['width'],
+				$image_meta['height'],
+			);
+		}
+	}
+
+	/**
+	 * Filters the image size for the image block.
+	 *
+	 * @since 4.5.0
+	 *
+	 * @param array $image_dimensions The calculated image size width and
+	 *                                height (in that order).
+	 * @param array $block_attributes The block attributes.
+	 * @param array $image_meta       The image attachment meta data.
+	 */
+	return apply_filters( 'block_core_image_get_width_height', $image_dimensions, $block_attributes, $image_meta );
+}
+
+/**
+ * Filters the rendered output of the Image block to include generated HTML
+ * attributes for front-end display.
+ *
+ * @since 4.5.0
+ *
+ * @param string $html  Original HTML.
+ * @param array  $block Parsed block.
+ * @return string Filtered Image block HTML.
+ */
+function gutenberg_render_block_core_image( $html, $block ) {
+	// Old post or... something's wrong.
+	if ( empty( $html ) || empty( $block['attrs'] ) ) {
```

Comment:
I think we should be shortcutting here if the block is not the image block? It'll eventually be shortcutted at the below test of `empty( $block_attributes['url'] )`, but it seems both indirect and wasteful to let us get that far, when this is called for every block.

Reply:
Right, good catch. Added in https://github.com/WordPress/gutenberg/pull/11973/commits/648811fd8c1bd3c62583c107928fada04d6af87e.